File: packages/template-python/src/deepagents_template/runtime/permissions.py
"""Permission policy resolution.

Port of ``src/runtime/permissions.ts``. Resolves the three permission modes
(``ask`` / ``yolo`` / ``plan``) into concrete allow/deny globs and HITL
interrupt-on maps.
"""

from __future__ import annotations

import fnmatch
from collections.abc import Iterable
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from deepagents_template.runtime.config.config_schema import (
    AppConfig,
    PermissionsConfig,
)


@dataclass
class SandboxPolicy:
    """Resolved sandbox policy — read/write allow/deny globs plus protected."""

    allowed_read_paths: list[str] = field(default_factory=list)
    allowed_write_paths: list[str] = field(default_factory=list)
    denied_read_paths: list[str] = field(default_factory=list)
    denied_write_paths: list[str] = field(default_factory=list)
    profile: str = "workspace"  # "open" | "workspace"


def resolve_sandbox_policy(
    config: AppConfig, workspace_root: Path | str | None = None
) -> SandboxPolicy:
    """Resolve a :class:`SandboxPolicy` for *config* against *workspace_root*."""
    permissions: PermissionsConfig = config.permissions

    allowed_read = list(permissions.allowed_paths or []) or ["/**"]
    allowed_write = list(permissions.allowed_paths or []) or ["/**"]
    denied_read = [str(p) for p in (permissions.denied_paths or [])]
    denied_write = [str(p) for p in (permissions.denied_paths or [])]

    profile = "open" if permissions.mode == "yolo" and not denied_write else "workspace"
    return SandboxPolicy(
        allowed_read_paths=allowed_read,
        allowed_write_paths=allowed_write,
        denied_read_paths=denied_read,
        denied_write_paths=denied_write,
        profile=profile,
    )


def to_absolute_deny_glob(denied: str, workspace_root: Path | str) -> str:
    """Return an absolute, glob-style deny entry — mirrors TS helper."""
    root = Path(workspace_root).expanduser().resolve()
    if denied.startswith("/"):
        return str(root / denied.lstrip("/")).rstrip("/") + "/**"
    return str((root / denied).resolve()) + "/**"


def build_permissions(
    config: AppConfig, workspace_root: Path | str | None = None
) -> list[dict[str, Any]]:
    """Translate :class:`PermissionsConfig` into deepagents-style permission dicts."""
    sandbox = resolve_sandbox_policy(config, workspace_root)
    root = Path(workspace_root).expanduser().resolve() if workspace_root else Path.cwd()
    rules: list[dict[str, Any]] = []

    if sandbox.allowed_write_paths:
        rules.append(
            {
                "operations": ["read", "write"],
                "paths": [str(root / p.lstrip("/")) for p in sandbox.allowed_write_paths],
                "mode": "allow",
            }
        )
    for denied in sandbox.denied_write_paths:
        rules.append(
            {
                "operations": ["write"],
                "paths": [to_absolute_deny_glob(denied, root)],
                "mode": "deny",
            }
        )
    return rules


def build_interrupt_on(tools: Iterable[str]) -> dict[str, bool]:
    """Build the ``interrupt_on`` dict consumed by deepagents/pydantic-ai."""
    return {tool: True for tool in tools}


def is_path_allowed(path: Path, globs: Iterable[str], workspace_root: Path) -> bool:
    """Return ``True`` if *path* matches any of *globs* (relative to *workspace_root*)."""
    target = path if path.is_absolute() else (workspace_root / path).resolve()
    for pattern in globs:
        absolute = pattern if pattern.startswith("/") else str(workspace_root / pattern)
        if fnmatch.fnmatch(str(target), absolute):
            return True
    return False

File: packages/deepagents-app-ts/src/runtime/permissions.ts
/**
 * Sandbox Policy & Filesystem Permissions
 *
 * Resolves the effective sandbox policy from config/profile and builds the
 * deepagents `FilesystemPermission[]` deny/allow rules. Deny entries are emitted
 * as both OS-absolute and workspace-relative globs so deepagents' built-in
 * FilesystemMiddleware enforces them regardless of the path form the model uses.
 */

import { join } from "node:path";
import { type FilesystemPermission } from "deepagents";
import { type AppConfig } from "./config/config-loader.js";

export interface SandboxPolicy {
  profile: "custom" | "workspace-write" | "read-only" | "open";
  deniedWritePaths: string[];
  writablePaths: string[];
  allowedEnv: string[];
  secretEnv: string[];
}

export function resolveSandboxPolicy(config: AppConfig): SandboxPolicy {
  const sandbox = config.sandbox;
  if (!sandbox || sandbox.profile === "custom") {
    return {
      profile: "custom",
      deniedWritePaths: config.permissions.deniedPaths,
      writablePaths: config.permissions.allowedPaths,
      allowedEnv: sandbox?.environment.allowedEnv ?? [],
      secretEnv: sandbox?.environment.secretEnv ?? [],
    };
  }

  if (sandbox.profile === "open") {
    return {
      profile: "open",
      deniedWritePaths: [],
      writablePaths: ["/**"],
      allowedEnv: sandbox.environment.allowedEnv,
      secretEnv: sandbox.environment.secretEnv,
    };
  }

  if (sandbox.profile === "read-only") {
    return {
      profile: "read-only",
      deniedWritePaths: ["/"],
      writablePaths: [],
      allowedEnv: sandbox.environment.allowedEnv,
      secretEnv: sandbox.environment.secretEnv,
    };
  }

  return {
    profile: "workspace-write",
    deniedWritePaths: sandbox.deniedWritePaths,
    writablePaths: sandbox.writablePaths,
    allowedEnv: sandbox.environment.allowedEnv,
    secretEnv: sandbox.environment.secretEnv,
  };
}

/**
 * Convert a single denied-write path (workspace-relative or absolute) into
 * the OS-absolute glob that deepagents' FilesystemMiddleware (decidePathAccess)
 * matches against the `file_path` the tool receives.
 *
 * Examples (with workspaceRoot = "/Users/dev/project"):
 *   "src/runtime/"     → "/Users/dev/project/src/runtime/**"
 *   "src/runtime"      → "/Users/dev/project/src/runtime/**"
 *   "/abs/dir/"        → "/abs/dir/**"
 *   "/"                → "/**"
 *
 * `join()` is used (not `resolve()`) so the trailing slash on a directory is
 * preserved — `/abs/dir/**` (correct) vs `/abs/dir**` (would also match
 * `/abs/dirfoo/x.ts`).
 */
export function toAbsoluteDenyGlob(denied: string, workspaceRoot: string): string {
  const withSlash = denied.endsWith("/") ? denied : `${denied}/`;
  const absoluteBase = withSlash.startsWith("/")
    ? withSlash
    : join(workspaceRoot, withSlash);        // join preserves trailing `/`
  return absoluteBase.startsWith("/") ? `${absoluteBase}**` : `/${absoluteBase}**`;
}

/**
 * Workspace-relative ("virtual") absolute deny glob — the path form the model
 * uses when addressing files relative to the backend root (e.g. "/src/x.ts").
 * Pairs with `toAbsoluteDenyGlob` so deny rules match whether deepagents'
 * FilesystemMiddleware receives an OS-absolute or backend-relative path.
 */
export function toWorkspaceDenyGlob(denied: string): string {
  const trimmed = denied.startsWith("/") ? denied : `/${denied}`;
  const withSlash = trimmed.endsWith("/") ? trimmed : `${trimmed}/`;
  return `${withSlash}**`;
}

/**
 * Build filesystem permissions for deepagents.
 *
 * Protects denied paths from writes while allowing everything else. The
 * deepagents `decidePathAccess` rule evaluator uses micromatch globs against
 * the literal `file_path` string the tool receives — which in our setup is an
 * OS-absolute path like `/Users/foo/project/src/runtime/x.ts`, not the
 * backend-rooted path `/src/runtime/x.ts`. So we resolve each denied entry
 * against the workspace root to produce an absolute glob before adding the
 * deny rule. Without this resolution, the deny glob never matches the actual
 * file path the agent passes and the deny is silently ignored.
 */
export function buildPermissions(config: AppConfig, workspaceRoot?: string): FilesystemPermission[] {
  const permissions: FilesystemPermission[] = [];
  const sandbox = resolveSandboxPolicy(config);
  const root = workspaceRoot ?? "/";

  for (const denied of sandbox.deniedWritePaths) {
    // Emit BOTH an OS-absolute glob and a workspace-relative ("virtual")
    // absolute glob. deepagents' FilesystemMiddleware matches the literal path
    // the model passes to the tool, which may be either form. Covering both
    // keeps the deny robust now that the custom protected-paths middleware
    // (which normalized paths itself) has been removed. Dedup so absolute
    // inputs (where both globs coincide) yield a single pattern.
    const denyGlobs = Array.from(
      new Set([toAbsoluteDenyGlob(denied, root), toWorkspaceDenyGlob(denied)])
    );
    permissions.push({
      operations: ["write"],
      paths: denyGlobs,
      mode: "deny" as const,
    });
  }

  permissions.push({
    operations: ["read", "write"],
    paths: ["/**"],
    mode: "allow" as const,
  });

  return permissions;
}

/**
 * Build the interruptOn map for deepagents from the config array.
 * Maps tool names to `true` for human-in-the-loop approval.
 */
export function buildInterruptOn(tools: string[]): Record<string, boolean> {
  const interruptOn: Record<string, boolean> = {};
  for (const toolName of tools) {
    interruptOn[toolName] = true;
  }
  return interruptOn;
}
